packages/app/src/cli/utilities/developer-platform-client/partners-client.test.ts

```diff
@@ -175,6 +185,81 @@ describe('fetchApp', async () => {
+    expect(partnersRequest).toHaveBeenCalledWith(FindOrganizationQuery, 'token', {id: ORG1.id}, undefined, undefined, {
+      type: 'token_refresh',
+      handler: expect.any(Function),
+    })
+  })
+})
+
+describe('Token refresh', () => {
+  let partnersClient: PartnersClient
+  let sessionMock: PartnersSession
+  const partnersRequestMock = vi.mocked(partnersRequest)
+  const query = `query { doesnt { matter } }`
+  const variables = undefined
+  const response = {}
+  const newToken = 'new-token'
+  let refreshedToken: string | undefined
+
+  beforeEach(() => {
+    sessionMock = testPartnersUserSession
+    partnersClient = new PartnersClient(sessionMock)
+    vi.spyOn(partnersClient, 'refreshToken').mockImplementation(vi.fn())
+    ;(partnersClient.refreshToken as Mock).mockResolvedValue(newToken)
+
+    partnersRequestMock.mockImplementation(
+      async (_query, _token, _variables, _headers, _preferredBehaviour, handler) => {
+        if (handler?.type === 'token_refresh') {
+          refreshedToken = (await handler.handler()).token
+        }
+        return Promise.resolve(response)
+      },
+    )
+  })
+
+  test('refreshes token once if the handler is called', async () => {
+    await expect(partnersClient.request(query, variables)).resolves.toEqual(response)
+
+    expect(refreshedToken).toBe(newToken)
+    expect(partnersClient.refreshToken).toHaveBeenCalledOnce()
+    expect(partnersRequestMock).toHaveBeenCalledTimes(1)
+  })
+
+  test('if token refresh is called twice, the token returns undefined', async () => {
+    partnersRequestMock.mockImplementation(
+      async (_query, _token, _variables, _headers, _preferredBehaviour, handler) => {
+        if (handler?.type === 'token_refresh') {
+          // Call handler twice, simulating a failure of the first call.
+          await handler.handler()
+          refreshedToken = (await handler.handler()).token
+        }
+        return Promise.resolve(response)
+      },
+    )
+
+    await expect(partnersClient.request(query, variables)).resolves.toEqual(response)
+
+    expect(refreshedToken).toBe(undefined)
+    expect(partnersClient.refreshToken).toHaveBeenCalledOnce()
+    expect(partnersRequestMock).toHaveBeenCalledTimes(1)
+  })
+
+  test('if multiple calls to partnerClient are made while token refresh is happening, the second call fails', async () => {
+    let refreshedToken: string | undefined
+    ;(partnersClient.refreshToken as Mock).mockImplementationOnce(async () => {
+      // Just sleep so the second call immediately trips the race condition guard clause
+      await new Promise((resolve) => setTimeout(resolve, 3000))
+      return newToken
+    })
+
+    const firstRequest = partnersClient.request(query, variables)
+    const secondRequest = partnersClient.request(query, variables)
+
+    await expect(secondRequest).rejects.toThrow('Multiple simultaneous token refresh attempts are not allowed')
+
+    expect(refreshedToken).toBe(undefined)
```

Comment:
shouldn't this be set? should we be awaiting the first request's (successful) conclusion?